alignment/batch_sequential_detector.py

```diff
@@ -0,0 +1,246 @@
+                    if re_ask_time == self.re_ask_times - 1:
+                        raise
+
+            with filename.open('w') as f: # 只有有用的output才会被cache
+                json.dump(output_text, f)
+        else:
+            with filename.open('r') as f:
+                output_text = json.load(f)
+                align_map, _, _ = GPTBatchSequentialDetector.lcs_sequence_alignment(raw_text, self.clearup_output(output_text))
+            
+        return align_map
+
+    def generate_batch(self, lines: list[str], begin_lineid: int) -> Tuple[str, int]:
+        """
+        从begin_lineid开始构造一个连续若干行的batch，使这个batch尽可能大，同时不超出self.token_limit指定的限制。
+
+        Args:
+            lines (list[str]): 源输入文件按行隔开的文本
+            begin_lineid (int): 欲开始构造的行下标，包含此行
+
+        Returns:
+            Tuple[str, int]: 构造的batch本身，以及游标处理到的下一行（即当前batch对应于原文的最后一行的*下一行*）行下标
+
+        """
+        assert begin_lineid < len(lines)
+        buffer = ''
+        for lineid in range(begin_lineid, len(lines)):
+            line = lines[lineid]
+            pending_text = (buffer + '\n' if len(buffer)>0 else '') + line
+            tokens = self.encoder.encode(pending_text)
+            if len(tokens) >= self.token_limit:
+                return buffer, lineid # 本行还没加上，所以是开区间
+            buffer = pending_text
+        if buffer:
+            return buffer, lineid + 1
+
+    def detect(self, lines: list[str], record_id: str, **kwargs) -> list[bool]:
+        """
+        Applies the GPT-3.5 detection technique to the given lines.
+        This method first batches the lines, processes the batches, and then post-processes the results.
+
+        Args:
+            lines (list[str]): The lines to be detected.
+            record_id (int): The unique id of the record. Use to cache the output of GPT
+
+        Returns:
+            list[bool]: The detection results.
+        """
+        # processed_batches = []
+        detections = [True] * (len(lines) - 1)
+
+        todo_lineid = 0
+        batch_id = 0
+
+        while todo_lineid < len(lines):
+            batch, next_line_id = self.generate_batch(lines, todo_lineid) # 利用已有的结果生成input_batch
+            if len(self.encoder.encode(batch)) < 20: # 结尾不能成段的噪声可能会让gpt疯狂道歉，这种情况下我们放过
+                break
+
+            align_map = self.align_gpt_linebreak_detection_request(batch, record_id, batch_id)
+            # Compare the hard line breaks in the raw text with the output text
+
+            input_line_offset = next_line_id - len(batch.splitlines()) # 第一行在本文件中的下标
+            assert lines[input_line_offset] == batch.splitlines()[0]
+            if next_line_id < len(lines):
+                todo_lineid = max(itertools.chain(*align_map.values())) + input_line_offset + 1 
+            else:
+                # 已经做完了本文件，接下来不再请求了
+                todo_lineid = len(lines)
+
+            for igroups in align_map.values():
+                for igroup in igroups:
+                    if igroup + 1 in igroups:
```

Comment:
Add doc string to explain why plus 1
Something 'the last value of a mapping is a hard linebreak'